feg/gateway/services/n7_n40_proxy/n7/model_conversion.go

```diff
@@ -0,0 +1,567 @@
+		policyRule := GetPolicyRuleProto(policyDecision, &pccRule)
+		if policyRule == nil {
+			// This is a static rule. Static rule only has pccRuleId and RefCondData
+			staticRule := &protos.StaticRuleInstall{
+				RuleId:           ruleId,
+				ActivationTime:   activationTime,
+				DeactivationTime: deactivationTime,
+			}
+			staticRules = append(staticRules, staticRule)
+			continue
+		}
+
+		dynamicRule := &protos.DynamicRuleInstall{
+			PolicyRule:       policyRule,
+			ActivationTime:   activationTime,
+			DeactivationTime: deactivationTime,
+		}
+		dynamicRules = append(dynamicRules, dynamicRule)
+	}
+
+	return staticRules, dynamicRules, rulesToRemove
+}
+
+func GetActivateDeactivationTime(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) (activationTime, deactivationTime *timestamp.Timestamp) {
+	if pccRule.RefCondData != nil && *pccRule.RefCondData != "" {
+		condData, found := policyDecision.Conds.Get(*pccRule.RefCondData)
+		if found {
+			activationTime = ConvertToProtoTimeStamp(condData.ActivationTime)
+			deactivationTime = ConvertToProtoTimeStamp(condData.DeactivationTime)
+		}
+	}
+	return
+}
+
+func GetPolicyRuleProto(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) *protos.PolicyRule {
+	// For all Ref*Data Maximum one entry is expected.
+	// These are declared as arrays for future compatibility
+	policyRule := &protos.PolicyRule{
+		Id:            pccRule.PccRuleId,
+		Priority:      getSbiUinteger(pccRule.Precedence),
+		MonitoringKey: getMonKeyFromPccRule(policyDecision, pccRule),
+		Redirect:      getTcDataFromPccRule(policyDecision, pccRule),
+		Qos:           getQosDataFromPccRule(policyDecision, pccRule),
+		FlowList:      getFlowListFromPccRule(pccRule),
+	}
+	if pccRule.RefChgData != nil && len(*pccRule.RefChgData) > 0 && policyDecision.ChgDecs != nil {
+		chgData, found := policyDecision.ChgDecs.Get((*pccRule.RefChgData)[0])
+		if found {
+			policyRule.RatingGroup = uint32(*chgData.RatingGroup)
+			policyRule.ServiceIdentifier = &protos.ServiceIdentifier{
+				Value: uint32(*chgData.ServiceId),
+			}
+			policyRule.Offline = getSbiBool(chgData.Offline)
+			policyRule.Online = getSbiBool(chgData.Online)
+		}
+	}
+	policyRule.TrackingType = getTrackingTypeProto(len(policyRule.MonitoringKey) != 0, policyRule.RatingGroup != 0)
+
+	if len(policyRule.MonitoringKey) == 0 && policyRule.Redirect == nil &&
+		policyRule.Qos == nil && len(policyRule.FlowList) == 0 && policyRule.RatingGroup == 0 {
+		// must be a static rule if these are not present
+		return nil
+	}
+	return policyRule
+}
+
+func getMonKeyFromPccRule(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) []byte {
+	if pccRule.RefUmData != nil && len(*pccRule.RefUmData) > 0 && policyDecision.UmDecs != nil {
+		umData, found := policyDecision.UmDecs.Get((*pccRule.RefUmData)[0])
+		if found {
+			return []byte(umData.UmId)
+		}
+	}
+	return []byte{}
+}
```

Comment:
shound't you be returning `nil`  instad of `[]byte{}`

Reply:
This value is returned to protobuf generated code which is of type []byte and not *[]byte.

Reply:
oh but slices are interfaces, meaning they can be nil. See [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Amagma%2Fmagma.git&branch=master&file=feg%2Fcloud%2Fgo%2Fprotos%2Fs8_proxy.pb.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.2.2&start_row=303&start_col=55&end_row=303&end_col=55) 

I think you should return nil here following the logic of the optional fields